server/src/main/java/org/eclipse/lsp/cobol/core/visitor/CobolVisitor.java

```diff
@@ -387,30 +387,34 @@ private void checkFileControlUniqueness(FileControlEntryContext ctx, String file
   @Override
   public List<Node> visitExecCicsStatement(ExecCicsStatementContext ctx) {
     areaBWarning(ctx);
-    EmbeddedCode code = embeddedCodeParts.get(ctx.cicsRules().getStart());
-    // apply area B check for tokens provided by a specific lexer
-    areaBWarning(code.getTokens());
-    return getLocality(ctx.getStart())
-        .<List<Node>>map(
-            it ->
-                ImmutableList.of(
-                    new EmbeddedCodeNode(
-                        it, code.getTokenStream(), code.getTree(), EmbeddedCodeNode.Language.CICS)))
-        .orElse(ImmutableList.of());
+    return processEmbeddedNodes(ctx, ctx.cicsRules(), EmbeddedCodeNode.Language.CICS);
   }
 
   @Override
   public List<Node> visitExecSqlStatement(ExecSqlStatementContext ctx) {
     areaBWarning(ctx);
-    EmbeddedCode code = embeddedCodeParts.get(ctx.sqlCode().getStart());
-    // apply area B check for tokens provided by a specific lexer
-    areaBWarning(code.getTokens());
-    return getLocality(ctx.getStart())
-        .<List<Node>>map(
-            it ->
-                ImmutableList.of(
-                    new EmbeddedCodeNode(
-                        it, code.getTokenStream(), code.getTree(), EmbeddedCodeNode.Language.SQL)))
+    return processEmbeddedNodes(ctx, ctx.sqlCode(), EmbeddedCodeNode.Language.SQL);
+  }
+
+  private List<Node> processEmbeddedNodes(ParserRuleContext parent,
+      ParserRuleContext ctx, EmbeddedCodeNode.Language language) {
+    return ofNullable(ctx)
+        .map(ParserRuleContext::getStart)
+        .map(embeddedCodeParts::get)
+        .map(
```

Comment:
Try to use `peek` here.

Reply:
The intention of `peek` is debugging, so we will have a code smell here.

Reply:
OK then.

Reply:
Still, it looks dirty, I will redo it.